src/rfxencode_diff_rlgr1.c

```diff
@@ -0,0 +1,281 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol client.
+ * RemoteFX Codec Library - RLGR
+ *
+ * Copyright 2011 Vic Lee
+ * Copyright 2016-2017 Jay Sorg <jay.sorg@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This implementation of RLGR refers to
+ * [MS-RDPRFX] 3.1.8.1.7.3 RLGR1/RLGR3 Pseudocode
+ */
+
+#if defined(HAVE_CONFIG_H)
+#include <config_ac.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "rfxcommon.h"
+
+#define GetNextInput do { \
+    input = *coef; \
+    coef++; \
+    coef_size--; \
+} while (0)
+
+#define CheckWrite do { \
+    while (bit_count >= 8) \
+    { \
+        bit_count -= 8; \
+        *cdata = bits >> bit_count; \
+        cdata++; \
+    } \
+} while (0)
+
+/*CodeGR(krp, lmag);*/ /* output GR code for (mag - 1) */
+#define CodeGR(_krp, _lmag) do { \
+    int lkr = _krp >> 3; \
+    /* unary part of GR code */ \
+    int lvk = _lmag >> lkr; \
+    /*OutputBit(lvk, 1);*/ \
+    int llvk = lvk; \
+    while (llvk >= 8) \
+    { \
+        bits <<= 8; \
+        bits |= 0xFF; \
+        llvk -= 8; \
+        *cdata = bits >> bit_count; \
+        cdata++; \
+    } \
+    bits <<= llvk; \
+    bits |= (1 << llvk) - 1; \
+    bit_count += llvk; \
+    /*OutputBit(1, 0);*/ \
+    bits <<= 1; \
+    bit_count++; \
+    CheckWrite; \
+    /* remainder part of GR code, if needed */ \
+    if (lkr) \
+    { \
+        /*OutputBits(lkr, _lmag & ((1 << lkr) - 1));*/ \
+        bits <<= lkr; \
+        bits |= _lmag & ((1 << lkr) - 1); \
+        bit_count += lkr; \
+    } \
+    /* update _krp, only if it is not equal to 1 */ \
+    if (lvk == 0) \
+    { \
+        /*UpdateParam(_krp, -2, lkr);*/ \
+        _krp -= 2; \
+        if (_krp < 0) \
+        { \
+            _krp = 0; \
+        } \
+    } \
+    else if (lvk > 1) \
+    { \
+        /*UpdateParam(_krp, lvk, lkr);*/ \
+        _krp += lvk; \
+        if (_krp > 80) \
+        { \
+            _krp = 80; \
+        } \
+    } \
+} while (0)
+
+int
+rfx_encode_diff_rlgr1(sint16 *coef, uint8 *cdata, int cdata_size)
+{
+    int k;
+    int kp;
+    int krp;
+
+    int input;
+    int numZeros;
+    int runmax;
+    int mag;
+    int sign;
+    int processed_size;
+    int lmag;
+    int coef_size;
+    int y;
+
+    int bit_count;
+    unsigned int bits;
+    uint8 *cdata_org;
+
+    uint32 twoMs;
+
+    for (k = 4095; k > 4032; k--)
```

Comment:
Please avoid magic numbers. Let's be explicit about limitations, assumptions and their interdependencies.